macros/timeHistoryAnalyzer.py

```diff
@@ -0,0 +1,457 @@
+                             data.mask[vfat,channel] != 0,
+                             maxSkip)
+
+    return list(filter(lambda r: r.maskedScanCount() >= minBadScans,
+                       ranges))
+
+def findRangesNoise(data,
+                    vfat,
+                    channel,
+                    threshold = 0.02,
+                    maxSkip = 5,
+                    minBadScans = 4):
+    """Finds ranges of scans based on noise.
+
+    Searches the data for the given vfat and channel for ranges of scans with
+    noise < threshold. During the search, at most maxSkip scans with mask not
+    set can be skipped. Only ranges with more than minBadScans are kept.
+
+    Args:
+        data: The TimeSeriesData object to pull data from
+        vfat: The VFAT to return ranges for
+        channel: The channel to return ranges for
+        threshold: The minimum value of noise for a scan to be "good"
+        maxSkip: The maximum number of "good" scans between two "bad" scans
+        minBadScans: The minimum number of bad scans
+
+    Returns:
+        A list of MaskedRange objects
+    """
+    ranges = _findRangesMeta(data,
+                             vfat,
+                             channel,
+                             data.noise[vfat,channel] < threshold,
+                             maxSkip)
+
+    return list(filter(lambda r: np.count_nonzero(r.noise() < threshold) >= minBadScans,
+                       ranges))
+
+class TimeSeriesData(object):
+    """Holds information about time variation of scan results.
+
+    Each property is stored as a 3D Numpy array with indexes
+    [vfat][strip][time]. The time index -> scan date mapping is exposed in the
+    date attribute.
+
+    Attributes:
+        dates: Array of strings containing the scan dates.
+        mask: Data for the "mask" property,
+        maskReason: Data for the "maskReason" property,
+    """
+
+    def __init__(self, inputDir):
+        """Creates a TimeSeriesData object by reading the files located in the
+        inputDir directory.
+
+        The input directory must contain the following files:
+
+        * gemPlotterOutput_mask_vs_scandate.root
+        * gemPlotterOutput_maskReason_vs_scandate.root
+
+        They are created by plotTimeSeries.py.
+
+        Args:
+            inputDir: The path to the input directory
+        """
+        import ROOT as r
+        from root_numpy import hist2array
+
+        file_mask = r.TFile('%s/gemPlotterOutput_mask_vs_scandate.root' % inputDir, 'READ')
+        file_maskReason = r.TFile('%s/gemPlotterOutput_maskReason_vs_scandate.root' % inputDir, 'READ')
+        file_noise = r.TFile('%s/gemPlotterOutput_noise_vs_scandate.root' % inputDir, 'READ')
+
+        self.mask = [] # [vfat][time][strip]; warning: reordered after loading
+        self.maskReason = [] # [vfat][time][strip]; warning: reordered after loading
+        self.noise = [] # [vfat][time][strip]; warning: reordered after loading
```

Comment:
Make these structured `numpy` arrays from the start, don't do gymnastics later to change them (lines 314-321).  You can then access them via their type dtype name then.

Reply:
> Why not make these structured `numpy` arrays from the start

Because then I can't use `hist2array` as easily. I have to do the gymnastics inside the loop or after.

> You can then access them via their type dtype name then

This isn't a list of structs, so I can't see how structured arrays would help (if that's what you're talking about).